### DISCUSSION TITLE
x-destroy like oposite function to x-init

### DISCUSSION THREAD

**Discussion by michalmatuska** [alpinejs/alpine#775] [4. General]
There are situations when you need add and initialize some external module with events and other logic things (like carousels or modals).
Now I have option to add some module with "onMount" method with x-init and works perfectly.
Problem come when I remove this mouted component. Everything from external module stay in memory and can cause problems.

**Comment by SimoTod** [comment]
It's been mentioned in the past but by design, Alpine tends to declare bahaviours as close as possible to the interaction that triggers them. 
X-init exists as a standalone directive because there's no user interaction at that point but destroy is often (always?) a consequence of your users doing something.
Most of the time you can "destroy" your carousel just before you remove the component following this pattern. 
```html
<div x-data="{init: () => externallibrary.init(...), destroy: () => {externallibrary.destroy(...); $el.remove(); }}" x-init="() => init()">
  ... 
  <button @click="destroy()">remove component</div>
</div>
```
Does it apply to your use case?

**Reply by michalmatuska** [reply]
Thanks for response.
Not exactly. When you remove something from the DOM it should run chaining reaction for call destroy function on all nested components. Every stable framework has this method and this is not accident. I understand. I should be close to declarative style of programming like it is possible but in some situations you can't.

**Reply by ryangjchandler** [reply]
@michalmatuska, I think what Simone is trying to say is that it doesn't really fit in my Alpine's philosophy, not that it's wrong to have something like `destroy`.

Also, not having a `destroy` callback / hook doesn't make Alpine un-stable as your implied in the message above.

**Reply by SimoTod** [reply]
For what it's worth, if your listener is not assigned to a variable and you remove the element from the DOM, the garbage collector will delete it eventually. I appreciate that there might be cases where you need to destroy an object but it's not as critical as in C.
It was a suggestion anyway, Caleb may find it appropriate so it's good to open a discussion.
At the moment, if you don't have control on what removes your component, the only solution is to implement a destroy function using a mutation observer but it's slightly more complex.

**Reply by michalmatuska** [reply]
@ryangjchandler Alpine has good philosophy. I like the idea of minimal framework and it is definitely something what I am looking for. But don't clean after myself can cause unpredictable issues. And this is something where is Alpine now unstable in my opinion. It can works i many cases. You can develop big parts of app with this. But once you cross this described line you will have problem.

@SimoTod Garbage collector is not working on everything. It is stronger mechanism than in past. But some cases are up to you. Like global changes - update style in head, add class to HTML node, attach listener to window etc. In this situations you need to do some manual work.

One example.
1) you install module like https://www.npmjs.com/package/scroll-lock
2) in some scenario you call its method `disablePageScroll` inside some Alpine component
3) then something (fetch and snippets update) change the DOM and it removes Alpine component
4) page isn't scrollable

And yes mutation observer is something what can help. What I know it is a part of Alpine now. I though that you have also some destroy mechanism for removing global events from x-on:event.window="" directive. It is something what can be exposed?

**Reply by SimoTod** [reply]
The current implementation removes event attached to the window object only when theybare called after the original component has been removed. They still stay in memory until that point so it's not what you need. 

If a developer use a specific package with global effect yeah, I agree, they should implement a way to restore the situation. In your example, what is actually removing the component? Couldn't the same function reenable the scroll?

I know that there's nothing forcing you to write that piece of code but there wouldn't be anything forcing you to write an x-destroy as well, it's just a different paradigm to achieve the same thing.

**Reply by michalmatuska** [reply]
Ok I understand. Very clever solution. Actually like all Alpine is.

I think, the force is DX, because these 2 reasons:

1) Some component did global change. Clear way (easy to read and understand) is do opposite action on same place. It is more predictable. This is reason why React has useEffect in this shape `useEffect(() => { init(); return destroy })`

2) DOM changes can happen from many places and actions. Yes you could do it like you described but it will not DRY and it is opened for future bugs. Not everybody (and future myself) will know what must be done for new action.
